KEP-2170: Add the TrainJob state transition design (#2298)
kubeflow/training-operator: docs/proposals/2170-kubeflow-training-v2/README.md

	Status TrainJobStatus `json:"status,omitempty"`
}

> const (

Reviewer: Can we define those const as `TrainJobConditionType` similar to Batch/Job: https://github.com/kubernetes/api/blob/master/batch/v1/types.go#L617 ?

Reviewer: Oh, you are right.

	// TrainJobSuspended means the TrainJob is suspended.
	TrainJobSuspended string = "Suspended"

> 	// TrainJobCompleted means that the actual jobs have completed its execution.

Reviewer: Should it be TrainJob, and similar for other statuses ?
```suggestion
	// TrainJobCompleted means that the TrainJob has completed its execution.
```

Author: Done.

> 	TrainJobComplete string = "Complete"

Author: We use "Complete" condition type name to align with batch/v1 Job for now, then we will open an issue to rename "Completed" to "Complete" in the JobSet side.

	// TrainJobFailed means that the actual jobs have failed its execution.
	TrainJobFailed string = "Failed"

	// TrainJobCreated means that the actual jobs creation has succeeded.
	TrainJobCreated string = "Created"
)

const (
	// TrainJobSuspendedReason is the "Suspended" condition reason.
	// When the TrainJob is suspended, this is added.
	TrainJobSuspendedReason string = "Suspended"

	// TrainJobResumedReason is the "Suspended" condition reason.
	// When the TrainJob suspension is changed from True to False, this is added.

> 	TrainJobResumedReason string = "Resumed"

Reviewer: @tenzen-y @kannon92 Do we have that status reason in JobSet or Job ?

Author: Yes, we have. In the batch/v1 Job level, "JobResumed" is used as a reason for the "Suspended" condition type.
In the JobSet level, "ResumeJobs" is used as a reason for the "Suspended" condition type.

	// TrainJobJobsCreationSucceededReason is the "Created" condition reason.
	// When the creating objects succeeded after building succeeded, this is added.
	TrainJobJobsCreationSucceededReason string = "JobsCreationSucceeded"

	// TrainJobJobsBuildFailedReason is the "Created" condition reason.
	// When the building objects based on the TrainJob and the specified runtime failed,
	// this is added.
	TrainJobJobsBuildFailedReason string = "JobsBuildFailed"

	// TrainJobJobsCreationFailedReason is "Created" condition reason.
	// When the creating objects failed even though building succeeded, this is added.

> 	TrainJobJobsCreationFailedReason string = "JobsCreationFailed"

Reviewer: Should we introduce those conditions in the 2nd iteration ?
I think, we should discuss if users really want to decouple conditions when TrainJob's object creation failed and when TrainJob fails.

Author: These are reasons, so we use these in the following. Additionally, if we do not distinguish `JobCreationFailed` and `JobBuildFailed`, it is hard to understand which part failed. So these reasons must be contained in the first phase.

- JobsBuildFailed: https://github.com/kubeflow/training-operator/blob/9e04bdd74920cbe12ecc786010d80fb8986f27bd/pkg/controller.v2/trainjob_controller.go#L71
- JobsCreationFailed: https://github.com/kubeflow/training-operator/blob/9e04bdd74920cbe12ecc786010d80fb8986f27bd/pkg/controller.v2/trainjob_controller.go#L98

```yaml
status:
  conditions:
  - type: Created
    status: false
    reason: JobsCreationFailed
```

```yaml
status:
  conditions:
  - type: Created
    status: false
    reason: JobsBuildFailed
```

```yaml
status:
  conditions:
  - type: Created
    status: true
    reason: JobsCreationSucceeded
```

Reviewer: Since we have validation webhook, what is the use-case you see when we can hit the JobsBuildFailed reason ?

Reviewer: So, I guess one of the examples could be if NewObject calls Kubernetes API server and this call fails.
In that case, we want to transition TrainJob to JobsBuildFailed status.

Author: > In that case, we want to transition TrainJob to JobsBuildFailed status.

Yes, that's right. In the current JobSet plugin, the building error could happen in the following:

- https://github.com/kubeflow/training-operator/blob/fd4d102f95af8c918f76d9d149dc3a6557d0512f/pkg/runtime.v2/framework/plugins/jobset/jobset.go#L93
- https://github.com/kubeflow/training-operator/blob/fd4d102f95af8c918f76d9d149dc3a6557d0512f/pkg/runtime.v2/framework/plugins/jobset/jobset.go#L98

Additionally, every plugin could return errors during Kubeflow Job Pipeline Framerowk.

            claimName: user-123-volume
```

### State Transition

In this section, we're explaining the TrainJob state transition (`.status.conditions`).
The basic TrainJob state machine is the below.
Especially, if we specify the TrainingRuntime or ClusterTrainingRuntime as a runtime,
the TrainJob terminal condition (`Failed` or `Completed`) is decided based on the JobSet terminal state (`status.terminalState`)
instead of computing from JobSet conditions.

```mermaid
stateDiagram-v2
    #CREATION
    state created_choice <<choice>>
    [*] --> created_choice: TrainJob is submitted.
    created_choice --> Created=True: Succeeded to build and deploy Jobs.

>     created_choice --> Created=False: Failed to build and deploy Jobs.

Reviewer: Would it be helpful to add small clarified message what does Failed to build Jobs mean ?

Author: Sure.

Author: Done.

>     Created=False --> Created=False: Wait for updated appropriate TrainJob.

Reviewer: What do you mean by Wait for updated appropriate TrainJob ?
I thought if build or deploy Jobs fails, we transition TrainJob to Failed condition ?

Author: TrainJob is the mutable object. So, we wait for updating TrainJob with proper fields here.

Reviewer: Some API in TrainJob is immutable, like managedBy, but container is mutable, right ?

Author: Yes, that's right. We allow users to modify the TrainJob except for some fields like managedBy.

Reviewer: Make sense, in that case my question is why do we transition TrainJob in the Failed state if the underlying JobSet is Failed ? Since TrainJob image is mutable, user can override it.
Or the main motivation is that when TrainJob is in Created state, it only can transition to Failed or Succeeded ?

Author: > in that case my question is why do we transition TrainJob in the Failed state if the underlying JobSet is Failed ? Since TrainJob image is mutable, user can override it.

AFAIK, both batch/v1 Job, and JobSet can not be restarted once those have reached the terminal phase (complete or failed).

    created_choice --> Created=True: Succeeded to build and deploy Jobs.
    created_choice --> Created=False: Failed to build and deploy Jobs.
    Created=False --> Created=False: Wait for updated appropriate TrainJob.
    Created=False --> Created=True: Succeeded to build and deploy Jobs.

    #SUSPENSION
    state suspended_choice <<choice>>
    Created=True --> suspended_choice: Handle TrainJob suspension.
    suspended_choice --> Suspended=True: TrainJob is suspended.
    Suspended=True --> Suspended=True: Wait for unsuspending.
    Suspended=True --> Suspended=False: TrainJob is unsuspended.
    suspended_choice --> Suspended=False: TrainJob is not suspended.

    #FAILURE
    state terminal_choice <<choice>>
    Suspended=False --> terminal_choice: Actual Jobs go to terminal phase.
    terminal_choice --> Failed=True: Actual Jobs (e.g., JobSet) failed.
    Failed=True --> [*]

    #COMPLETION
    terminal_choice --> Completed=True: Actual Jobs (e.g., JobSet) completed.
    Completed=True --> [*]
```

Hence, when we specify the JobSet as a runtime, after the TrainJob completed,
we can see the TrainJob pseudo state in the following:

```yaml
[...]
status:
  conditions:

>   - type: Created

Reviewer 2: Would it be helpful to add transition time and also reason for failure within conditions? I believe it would be help debugging failures

Author: This is pseudo conditions. So, actually we support all fields supported in the metav1.Conditions as you can see there: https://github.com/kubeflow/training-operator/blob/66ebecd12e689ffccc5b050560bd370c1bcc125d/pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go#L261-L262

    status: true
  - type: Suspended
    status: false
  - type: JobSetSuspended
    status: false
  - type: JobSetCompleted

>     status: true

Reviewer: Why do we need to propagate JobSet conditions to the TrainJob ?

Author: I propagated these JobSet conditions so that runtimes can be propagated to the arbitrary conditions.
But, in the JobSet case, these seem to be slightly redundant. So, I'm wondering if we only extend the runtime interface so that the plugin can propagate arbitrary conditions, but JobSet plugin does not propagate own conditions, then TrainJob does not have the JobSet conditions.

But, I guess that the JobSet StartupPolocy conditions could help to understand what happens in the actual Jobs.
So, we can propagate only `StartupPolicyInProgress` and `StartupPolicyCompleted`. 

https://github.com/kubernetes-sigs/jobset/blob/440f53e13ed1db15d4f1d5a04c2450e74df4d1e8/api/jobset/v1alpha2/jobset_types.go#L75-L78

@andreyvelich WDYT?

Reviewer: >So, I'm wondering if we only extend the runtime interface so that the plugin can propagate arbitrary conditions, but JobSet plugin does not propagate own conditions, then TrainJob does not have the JobSet conditions.

I would suggest that we add those once we have initial implementation of TrainJob status.
I think, this is valid option, but we should discuss what conditions be be propagated by runtime plugins.

Any thoughts @tenzen-y ?

Author: > I would suggest that we add those once we have initial implementation of TrainJob status.
I think, this is valid option, but we should discuss what conditions be be propagated by runtime plugins.

I think that we need to implement the conditions propagation mechanism, but it's not clear in the JobSet situation. Especially, which conditions should be propagated to the TrainJob. So, I would propose that we extend the runtime interface so that plugins can propagate the conditions to TrainJob. But, in the first status implementation, we do not propagate any JobSet conditions to the TrainJob.

What do you think? @andreyvelich

Reviewer: >So, I would propose that we extend the runtime interface so that plugins can propagate the conditions to TrainJob. 

Sure, we can have this. Maybe we should create a tracking issue to implement it ?
I don't think is has a high priority compare to other tasks.

Author: Yes, sure. I will open the issue.
But, let me mention the propagation mechanism in this KEP.

Author: Done.

instead of computing from JobSet conditions.

```mermaid
stateDiagram-v2
    #CREATION
    state created_choice <<choice>>
    [*] --> created_choice: TrainJob is submitted.
    created_choice --> Created=True: Succeeded to build and deploy Jobs.
    created_choice --> Created=False: Failed to build and deploy Jobs.
    Created=False --> Created=False: Wait for updated appropriate TrainJob.
    Created=False --> Created=True: Succeeded to build and deploy Jobs.

    #SUSPENSION
    state suspended_choice <<choice>>
    Created=True --> suspended_choice: Handle TrainJob suspension.
    suspended_choice --> Suspended=True: TrainJob is suspended.
    Suspended=True --> Suspended=True: Wait for unsuspending.
    Suspended=True --> Suspended=False: TrainJob is unsuspended.
    suspended_choice --> Suspended=False: TrainJob is not suspended.

    #FAILURE
    state terminal_choice <<choice>>
    Suspended=False --> terminal_choice: Actual Jobs go to terminal phase.
    terminal_choice --> Failed=True: Actual Jobs (e.g., JobSet) failed.
    Failed=True --> [*]

    #COMPLETION

>     terminal_choice --> Completed=True: Actual Jobs (e.g., JobSet) completed.

Reviewer: Should we actually call this condition Succeeded to be consistent with JobSet: https://github.com/kubernetes-sigs/jobset/blob/main/api/jobset/v1alpha2/jobset_types.go#L182-L183 ?

From my understanding Completed: True mens that TrainJob is in Failed or Succeeded state.

Author: Basically, Job and JobSet succeed (.replicatedJobsStatus.succeed), and failed (.replicatedJobsStatus.failed) count are not terminal phase, and those are intermediate one. Additionally those are not guaranteed consistency when Success and Failure criteria are conflicts.

So, instead of success count, we should rely on the .status. terminal state.

Author: TrainJob has only Suspend, Completed, Failed, and Created conditions.
